### ISSUE TITLE
write after end

### ISSUE THREAD

**Issue by kilianc** [felixge/node-stream-cache#6]
I took a quick look at the implementation and it looks like all `Stream#write()` calls are used as if they were synchronous, but [they are not](https://nodejs.org/dist/latest-v8.x/docs/api/stream.html#stream_writable_write_chunk_encoding_callback).

The result is that we close the stream before all writes are done.

Am I correct? Or there is something that I am missing?

**Comment by mariomurrent-softwaresolutions** [comment]
You're totally right.

**Comment by felixge** [comment]
@kilianc is there a bug/problem you have related to this? IIRC the `write` method in this pkg doesn't implement the callback argument because it isn't quite clear how it should behave, especially when multiple destinations are being piped to.

**Comment by kilianc** [comment]
@felixge yeah, I put this in my rev-proxy to re-attempt calls with a body and it fails with > 1MB multipart payloads.

**Comment by felixge** [comment]
@kilianc why is it failing? Is your lib expecting write() to provide a callback?

**Comment by kilianc** [comment]
@felixge it's a simple pipe. All I get is `write after end` error from the stream, see the code:

```js
pipe(sourceRequest, cache, targetRequest).on('error', reject)
```

This runs in a loop if there is a connection/dns/socket/etc error to target request, but the second time fails if the payload is big enough to need a callback.

**Comment by Ashoat** [comment]
Does anybody know of an alternative that doesn't have the same issues?

**Comment by Ashoat** [comment]
Following up on this in case it's useful for anybody else: we ended up [switching](https://github.com/CommE2E/comm/commit/76dadef030be10671e16b31391044435c87edf1f) to [`rereadable-stream`](https://github.com/signicode/rereadable-stream) and it solved the issue for us.
